### PR TITLE
Steckler

### PR DESCRIPTION
**Pull request by LinhWuerzburger**

### CODE REVIEW

**Comment by LinhWuerzburger** [comment]
steckler case does not works with multigrid level > 1.

**Comment by LinhWuerzburger** [comment]
I assume it is because of merged obstacles in the coarse levels

**Comment by LinhWuerzburger** [comment]
differ between obstacle and effectiv obstacle boundary list?

**Comment by LinhWuerzburger** [comment]
at the moment boundary cells of obstacles which cause circular constraints are removed, but only on the finest level. Is it necessary or relevant to repeat the process on the coarser level as well?

File: src/diffusion/ExplicitDiffuse.cpp [hunk 1]
```diff
@@ -138,18 +142,26 @@ void ExplicitDiffuse::ExplicitStep(Field *out, const Field *in, const real D, co
...
-                           + nu_z1 * (d_in[idx] - d_in[IX(i, j, k - 1, Nx, Ny)]) - nu_z2 * (d_in[IX(i, j, k + 1, Nx, Ny)] - d_in[idx])
+        real nu_x1 = 0.5 * (d_ev[IX(i - 1, j, k, Nx, Ny)] + d_ev[idx]) + D;
+        real nu_x2 = 0.5 * (d_ev[IX(i + 1, j, k, Nx, Ny)] + d_ev[idx]) + D;
+        real nu_y1 = 0.5 * (d_ev[IX(i, j - 1, k, Nx, Ny)] + d_ev[idx]) + D;
+        real nu_y2 = 0.5 * (d_ev[IX(i, j + 1, k, Nx, Ny)] + d_ev[idx]) + D;
+        real nu_z1 = 0.5 * (d_ev[IX(i, j, k - 1, Nx, Ny)] + d_ev[idx]) + D;
+        real nu_z2 = 0.5 * (d_ev[IX(i, j, k + 1, Nx, Ny)] + d_ev[idx]) + D;
+
+        auto di0 = (d_in[idx] - d_in[IX(i - 1, j, k, Nx, Ny)]); // * 4 / dx;  // positiv wenn nachbarzelle niedriger // u_i - u_{i-1}
+        auto di1 = (d_in[IX(i + 1, j, k, Nx, Ny)] - d_in[idx]); // positiv wenn nachbarzelle hoeher // u_{i+1} - u_i
```
Comment (ChristianVonMach) [inline src/diffusion/ExplicitDiffuse.cpp:153]: we are using english comments

File: src/utility/Utility.cpp [hunk 2]
```diff
@@ -114,23 +80,134 @@ std::shared_ptr<spdlog::logger> create_logger(std::string logger_name) {
...
+    }
+    // size_t Nx = Domain::getInstance()->get_Nx();
+    // size_t Ny = Domain::getInstance()->get_Ny();
+    // std::string index;
+    // for (size_t idx: indices) {
+    //     size_t k = getCoordinateK(idx, Nx, Ny);
+    //     size_t j = getCoordinateJ(idx, Nx, Ny, k);
+    //     size_t i = getCoordinateI(idx, Nx, Ny, j, k);
+    //     index += " (" + std::to_string(i) + "|" + std::to_string(j) + "|" + std::to_string(k) + ")";
+    // }
```
Comment (ChristianVonMach) [inline src/utility/Utility.cpp:188]: is this code not longer needed -> delete

File: src/boundary/Obstacle.cpp [hunk 3]
```diff
@@ -106,8 +90,8 @@ void Obstacle::init(size_t level) {
     m_obstacleLeft = new size_t[m_size_obstacleLeft];
     m_obstacleRight = new size_t[m_size_obstacleRight];
 
-    //m_size_obstacleInner = (m_strideX - 2) * (m_strideY - 2) * (m_strideZ - 2);
-    //m_obstacleInner = new size_t[m_size_obstacleInner];
+    // m_size_obstacleInner = (m_strideX - 2) * (m_strideY - 2) * (m_strideZ - 2);
+    // m_obstacleInner = new size_t[m_size_obstacleInner];
```
Comment (ChristianVonMach) [inline src/boundary/Obstacle.cpp:94]: is this code not longer needed -> delete

File: src/boundary/Obstacle.cpp [hunk 4]
```diff
@@ -210,18 +194,18 @@ void Obstacle::createObstacle(size_t Nx, size_t Ny) {
...
-    //            size_t idx = IX(i, j, k, strideX, strideY);
-    //            *(m_obstacleInner + index) = m_obstacleList[idx];
-    //        }
-    //    }
-    //}
+    //             size_t idx = IX(i, j, k, strideX, strideY);
+    //             *(m_obstacleInner + index) = m_obstacle_list[idx];
+    //         }
+    //     }
+    // }
```
Comment (ChristianVonMach) [inline src/boundary/Obstacle.cpp:208]: is this code not longer needed -> delete

**Review by ChristianVonMach — CHANGES_REQUESTED** [review]

**Comment by LinhWuerzburger** [comment]
the code is uncommented because currently it is not used. though we are not sure if we are going to need the code that is why it is still there. Shall I delete it nonetheless?

File: src/diffusion/ExplicitDiffuse.cpp (hunk 1 above)
Comment (LinhWuerzburger) [reply src/diffusion/ExplicitDiffuse.cpp]: done in cf8f624165ff0537415b34573befef053466d636

**Comment by ChristianVonMach** [comment]
> the code is commented because currently it is not used. though we are not sure if we are going to need the code that is why it is still there. Shall I delete it nonetheless?

we should not maintain code not in used and make our files bigger than needed. If we need this code in the future you can check it out via git.

File: src/boundary/Obstacle.cpp (hunk 4 above)
Comment (LinhWuerzburger) [reply src/boundary/Obstacle.cpp]: 5c97868f4303b2d83989a6c476cfcb68930c7c12

File: src/boundary/Obstacle.cpp [hunk 5]
```diff
@@ -122,7 +106,7 @@ Obstacle::~Obstacle() {
     delete (m_obstacleBottom);
     delete (m_obstacleLeft);
     delete (m_obstacleRight);
-    //delete (m_obstacleInner);
+    // delete (m_obstacleInner);
```
[the remark this answers is not in the document]
Comment (LinhWuerzburger) [reply src/boundary/Obstacle.cpp]: 5c97868f4303b2d83989a6c476cfcb68930c7c12

File: src/boundary/Obstacle.cpp (hunk 3 above)
Comment (LinhWuerzburger) [reply src/boundary/Obstacle.cpp:94]: 5c97868f4303b2d83989a6c476cfcb68930c7c12

**Review by ChristianVonMach — DISMISSED** [review]

**Review by ChristianVonMach — APPROVED** [review]

**Pull request merged**